verilog/tools/kythe/indexing_facts_tree_extractor.h

```diff
@@ -0,0 +1,83 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_TOOLS_KYTHE_INDEXING_FACTS_TREE_EXTRACTOR_H_
+#define VERIBLE_VERILOG_TOOLS_KYTHE_INDEXING_FACTS_TREE_EXTRACTOR_H_
+
+#include "absl/strings/string_view.h"
+#include "common/text/tree_context_visitor.h"
+#include "common/text/tree_utils.h"
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/tools/kythe/indexing_facts_tree.h"
+
+namespace verilog {
+namespace kythe {
+
+// Type that is used to keep track of the path to the root of indexing facts
+// tree.
+using IndexingFactsTreeContext = std::vector<IndexingFactNode*>;
+
+// This class is used for traversing CST and extracting different indexing
+// facts from CST nodes and constructs a tree of indexing facts.
+class IndexingFactsTreeExtractor : public verible::TreeContextVisitor {
+ public:
+  IndexingFactsTreeExtractor(absl::string_view base,
+                             absl::string_view file_name)
+      : context_(verible::TokenInfo::Context(base)) {
+    facts_tree_context_.push_back(&root_);
+  }
+
+  void Visit(const verible::SyntaxTreeNode& node) override;
+
+  const IndexingFactNode& GetRoot() { return root_; }
+
+ private:
+  // Extracts modules and creates its corresponding fact tree.
+  void ExtractModule(const verible::SyntaxTreeNode& node);
+
+  // Extracts modules instantiations and creates its corresponding fact tree.
+  void ExtractModuleInstantiation(const verible::SyntaxTreeNode& node);
+
+  // Extracts endmodule and creates its corresponding fact tree.
+  void ExtractModuleEnd(const verible::SyntaxTreeNode& node);
+
+  // Extracts modules headers and creates its corresponding fact tree.
+  void ExtractModuleHeader(const verible::SyntaxTreeNode& node);
+
+  // The Root of the constructed tree
+  IndexingFactNode root_ =
+      IndexingFactNode(IndexingNodeData(IndexingFactType::kFile));
```

Comment:
Since this is a non-trivial initialization, I would move this into the constructor definition.

Reply:
I couldn't  move it 'cos of the the default constructor is deleted in VectorTree.